### PR TITLE
wire contexts into bitswap requests more deeply

### PR DESCRIPTION
**Pull request by whyrusleeping**
This should address the issue of wantlist entries persisting in the wantlist and not being cancelled when things like 'ipfs get' are killed early

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

File: exchange/bitswap/bitswap.go [hunk 1]
```diff
@@ -208,11 +208,14 @@ func (bs *Bitswap) GetBlocks(ctx context.Context, keys []key.Key) (<-chan *block
 		log.Event(ctx, "Bitswap.GetBlockRequest.Start", &k)
 	}
 
-	bs.wm.WantBlocks(keys)
+	bs.wm.WantBlocks(ctx, keys)
 
-	req := &blockRequest{
-		keys: keys,
-		ctx:  ctx,
+	// NB: Optimization. Assumes that providers of key[0] are likely to
```
Comment (rht) [inline exchange/bitswap/bitswap.go]: Yeah, I thought broadcasting every blocks is too granular considering the bandwidth cost & rtt. This reduces dup subsequent blocks download as well?

**Comment by rht** [comment]
Tested
- `ipfs daemon &`
- `ipfs bitswap stat` no wantlist
- `ipfs get $sintel`
- `ipfs bitswap stat` in another terminal, contains the hash
- C-c `ipfs get`
- `ipfs bitswap stat` # wantlist contains the hash
- `ipfs bitswap sta`t #>10s later wantlist still contains the hash

**Comment by rht** [comment]
I tried catching the `ctx.Done()` to print the message, inside the `Run` phase of `ipfs get`.

The msg only gets printed when the daemon exits, but not when the client exits.
This would mean any pending request by a client won't ever be canceled?

**Comment by whyrusleeping** [comment]
@rht, thanks for investigating. I dont think the context from the api request is being passed down to the getblocks requst properly. I need to look at that wiring a little more closely.

**Comment by whyrusleeping** [comment]
it looks like closenotify isnt working for some reason... not sure why

**Comment by rht** [comment]
Not sure the reason with why another transport is created just to cancel the client request [here](https://github.com/ipfs/go-ipfs/blob/master/commands/http/client.go#L131).

When I replaced it with the transport created [previously](https://github.com/ipfs/go-ipfs/blob/master/commands/http/client.go#L43), I was able to C-c the client once. Currently it has to be done twice. Is this a bug or feature?

**Comment by rht** [comment]
test: I put a 1s timeout on the http handler that auto-cancel the connection. The wantlist still doesn't get gc-ed.

File: exchange/bitswap/bitswap.go (hunk 1 above)
Comment (rht) [inline exchange/bitswap/bitswap.go]: I think this should be separated out into another PR.

**Comment by whyrusleeping** [comment]
this is waiting until we sort out the CloseNotify nonsense.

**Comment by rht** [comment]
I tried simulating the closenotify, but in this PR, the wantlist still doesn't get gc-ed.
It does, however, in #1995.

**Comment by jbenet** [comment]
would be awesome to get this figured out-- am sure lots of traffic happens because of these lingering ctxs
